src/wallet/Wallet.ts

```diff
@@ -1680,22 +1662,45 @@ class Wallet {
 	/**
 	 * @deprecated Use createMeltQuoteBolt11.
 	 */
-	async createMeltQuote(invoice: string): Promise<MeltQuoteBolt11Response> {
-		return this.createMeltQuoteBolt11(invoice);
+	async createMeltQuote(invoice: string, amountMsat?: number): Promise<MeltQuoteBolt11Response> {
+		return this.createMeltQuoteBolt11(invoice, amountMsat);
 	}
 
 	/**
 	 * Requests a melt quote from the mint. Response returns amount and fees for a given unit in order
 	 * to pay a Lightning invoice.
 	 *
 	 * @param invoice LN invoice that needs to get a fee estimate.
+	 * @param amountMsat Optional amount in millisatoshis to attach for amountless invoices, must not
+	 *   be provided for invoices that already encode an amount.
 	 * @returns The mint will create and return a melt quote for the invoice with an amount and fee
 	 *   reserve.
 	 */
-	async createMeltQuoteBolt11(invoice: string): Promise<MeltQuoteBolt11Response> {
+	async createMeltQuoteBolt11(
+		invoice: string,
+		amountMsat?: number,
+	): Promise<MeltQuoteBolt11Response> {
+		if (amountMsat != null && invoiceHasAmountInHRP(invoice)) {
```

Comment:
If `amountMsat` is not passed (evaluating to `undefined`) this branch still triggers. Instead it should check whether `amountMsat` is defined and then use the `assertAmount` util as @zk-bits mentioned.

Reply:
The check is loose equality `!=`, so although it looks like it shouldn't pass, it will, because `undefined` and `null` are equivalent in loose equality. So it's technically correct, but would be clearer as:

```
		if (amountMsat != undefined && invoiceHasAmountInHRP(invoice)) {
```

That said, amount `0` is a nonsense input... so would should decide whether to throw or silently discard it. If we change the test to a truthy one, it will catch `0` and throw the custom error.

```suggestion
		if (amountMsat && invoiceHasAmountInHRP(invoice)) {
```

Then, if we do the `assertAmount()` straight after as @zk-bits  suggests, it will throw the general assert for non-integer, negative and other nonsense.

Reply:
In fact, as we are in Wallet, you can lean on the `this.failIf()` which does logging plus throw:

```ts
if (amountMsat) {
	this.failIf(invoiceHasAmountInHRP(invoice), 'amountMsat supplied but invoice already contains an amount. Leave amountMsat undefined for non-zero invoices.');
	assertAmount(amountMsat);
}
```